File: Tease AI/Services/CldAccessor.vb
Option Strict On
Option Explicit On
Imports System.IO
Imports TeaseAI.Common
Imports TeaseAI.Common.Data
Imports TeaseAI.Common.Interfaces.Accessors

Public Class CldAccessor
    Implements ICldAccessor

    Public Sub WriteCld(cldData As List(Of ScriptMetaData), filePath As String) Implements ICldAccessor.WriteCld
        If Not Directory.Exists(Path.GetDirectoryName(filePath)) Then
            Directory.CreateDirectory(Path.GetDirectoryName(filePath))
        End If

        Using fs As New FileStream(filePath, IO.FileMode.Create), BinWrite As New BinaryWriter(fs)
            For i = 0 To cldData.Count - 1
                BinWrite.Write(cldData(i).Name)
                BinWrite.Write(cldData(i).IsEnabled)
            Next
        End Using
    End Sub

    Public Function ReadCld(scriptHomeDir As String, fileName As String) As Result(Of List(Of ScriptMetaData)) Implements ICldAccessor.ReadCld
        If Not File.Exists(fileName) Then
            Return Result.Fail(Of List(Of ScriptMetaData))("Checklist file " + fileName + " does not exist")
        End If
        Dim returnValue As List(Of ScriptMetaData) = New List(Of ScriptMetaData)()
        Using fs As New FileStream(fileName, FileMode.Open), binRead As New BinaryReader(fs)
            Do While fs.Position < fs.Length
                Dim name As String = binRead.ReadString()
                Dim cldData As ScriptMetaData = New ScriptMetaData With {
                    .Key = scriptHomeDir & Path.DirectorySeparatorChar & name,
                    .Name = name,
                    .IsEnabled = binRead.ReadBoolean(),
                    .Info = String.Empty
                }
                returnValue.Add(cldData)
            Loop
        End Using
        Return Result.Ok(returnValue)
    End Function

End Class

File: TeaseAI.Services/Accessors/CldAccessor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using TeaseAI.Common;
using TeaseAI.Common.Data;
using TeaseAI.Common.Interfaces.Accessors;

namespace TeaseAI.Services.Accessors
{
    public class CldAccessor : ICldAccessor
    {
        public Result<List<ScriptMetaData>> ReadCld(string fileName)
        {
            var data = Result.Ok()
                .Ensure(() => File.Exists(fileName), "Checklist file " + fileName + " does not exist")
                .OnSuccess(() =>
                {
                    using (var fs = new FileStream(fileName, FileMode.Open))
                    using (var binRead = new BinaryReader(fs))
                    {
                        var returnValue = new List<ScriptMetaData>();
                        while (fs.Position < fs.Length)
                        {
                            var cldData = new ScriptMetaData
                            {
                                Key = String.Empty,
                                Name = binRead.ReadString(),
                                IsEnabled = binRead.ReadBoolean(),
                                Info = String.Empty
                            };
                            returnValue.Add(cldData);
                        }
                        return Result.Ok(returnValue);
                    }
                });
            return data;
        }

        public Result<List<ScriptMetaData>> ReadCld(string scriptHomeDir, string fileName)
        {
            var data = Result.Ok()
                .Ensure(() => File.Exists(fileName), "Checklist file " + fileName + " does not exist")
                .OnSuccess(() =>
                {
                    using (var fs = new FileStream(fileName, FileMode.Open))
                    using (var binRead = new BinaryReader(fs))
                    {
                        var returnValue = ReadDirectory(scriptHomeDir);
                        while (fs.Position + 1 < fs.Length)
                        {
                            var name = binRead.ReadString();
                            var key = scriptHomeDir + name + ".txt";
                            var cldData = returnValue.FirstOrDefault(smd => smd.Key == key);
                            if (cldData == null)
                            {
                                cldData = new ScriptMetaData { Key = key };
                                returnValue.Add(cldData);
                            }
                            cldData.Key = key;
                            cldData.Name = name;
                            cldData.IsEnabled = cldData.IsEnabled && binRead.ReadBoolean() && File.Exists(key);
                            cldData.Info = string.Empty;
                            cldData.IsChastity = name.ToLower().Contains("_chastity");
                            cldData.IsBeg = name.ToLower().Contains("_beg");
                            cldData.IsEdge = name.ToLower().Contains("_edge");
                            cldData.IsRestricted = name.ToLower().Contains("_restricted");
                            //returnValue.Add(cldData);
                        }
                        return Result.Ok(returnValue);
                    }
                });
            return data;
        }

        public void WriteCld(List<ScriptMetaData> cldData, string filePath)
        {
            Directory.CreateDirectory(Path.GetDirectoryName(filePath));
            using (var fs = new FileStream(filePath, FileMode.Create))
            using (var binWrite = new BinaryWriter(fs))
            {
                foreach (var cld in cldData)
                {
                    binWrite.Write(cld.Name);
                    binWrite.Write(cld.IsEnabled);
                }
            }
        }

        private List<ScriptMetaData> ReadDirectory(string scriptHomeDir)
        {
            var files = Directory.GetFiles(scriptHomeDir);
            return files.Select(f => new ScriptMetaData
            {
                Key = f,
                Name = Path.GetFileNameWithoutExtension(f),
                IsEnabled = true,
                Info = string.Empty,
            }).ToList();
        }
    }
}
